cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
   // return array_contains(hash_aggregations, t);
   return (t == aggregation::SUM) or (t == aggregation::MIN) or
          (t == aggregation::MAX) or (t == aggregation::COUNT) or
-         (t == aggregation::MEAN);
+         (t == aggregation::ARGMIN) or (t == aggregation::ARGMAX);
+}
+
+// flatten aggs to filter in single pass aggs 
+std::tuple<table_view, std::vector<aggregation::Kind>, std::vector<size_t> >
+flatten_single_pass_aggs(std::vector<aggregation_request> const& requests) {
+  std::vector<column_view> columns;
+  std::vector<aggregation::Kind> agg_kinds;
+  std::vector<size_t> col_ids;
+
+  for (size_t i = 0; i < requests.size(); i++) {
+    auto const& request = requests[i];
+    auto const& agg_v = request.aggregations;
+
+    auto insert_agg = [&agg_kinds, &columns, &col_ids, &request, i]
+    (aggregation::Kind k) {
+      agg_kinds.push_back(k);
+      columns.push_back(request.values);
+      col_ids.push_back(i);
+    };
+
+    std::for_each(agg_v.begin(), agg_v.end(),
+      [&columns, &agg_kinds, &request, &col_ids, insert_agg] 
+      (std::unique_ptr<aggregation> const& agg) {
+
+        if (is_hash_aggregation(agg->kind)) {
+          if (is_fixed_width(request.values.type())) {
+            insert_agg(agg->kind);
+          } else if (request.values.type().id() == type_id::STRING) {
```

Comment:
I've wanted to do this for so long. This would've been useful in sort groupby which also makes heavy use of dependency aggs. But I've not been able to come up with an infrastructure quite generic enough that it allows for all the various ways in which aggs depend on other aggs. e.g. in `nth` implemented in #4172, the count to use depends on aggregation parameter. https://github.com/rapidsai/cudf/blob/9d1482e23e72a35b6377835c7c1f7477524e9145/cpp/src/groupby/sort/groupby.cu#L326

Reply:
Agreed, it is tricky. Lets table it for future work and we can come up with something nice.